microsoft/dicom-server
# JWT Authentication 2 - Add Development Identity Provider

## Description
Follows on from [this PR](https://github.com/microsoft/dicom-server/pull/21).
Adds a DevelopmentIdentityProvider (in-memory identity server) with `TestAuthEnvironment.json` based on the FHIR Server equivalent which is [documented here](https://github.com/microsoft/fhir-server/blob/master/docs/Authentication.md).

## Important
As mentioned in the previous authentication PR, the Authority needs to be over https. Since the authority is now an identity server which is part of the Dicom Server, you need to make sure the Dicom Server is using https.
To do this enable SSL in the `Microsoft.Health.Dicom.Web` Debug Properties.

## Testing
Note that (in the testing example in the last PR) where we could use an external identity server such as the Fhir Server's DevelopmentIdentityProvider to create the JWT tokens, now we cannot. This is because the audience has been set to "dicom-api" and the Fhir Server will only produce tokens for "fhir-api". To test this with an external identity server consider [using AAD](https://blog.jongallant.com/2017/03/azure-active-directory-access-tokens-postman/).

To test using the DevelopmentIdentityProvider which this PR adds:
1. Enable SSL in the `Microsoft.Health.Dicom.Web` Debug Properties.
2. Enable Security in `appsettings.json` and start the server. Make sure the server is running at `https://localhost:44348` (this should be the Audience in `appsettings.json`)
3. Send the following request to the server to get a `access_token` in the response:
```
POST /connect/token HTTP/1.1
Host: https://localhost:44348
Content-Type: application/x-www-form-urlencoded

client_id=serviceclient&client_secret=serviceclient&grant_type=client_credentials&scope=fhir-api
```
4. Then use any endpoint on the Dicom Server by providing the token in a Authorization header. For example:
```
Authorization: Bearer eyJhbGciOiJSUzI1NiIsImtpZCI6Ijc4YWJlMDM0OGEyNDg4NzU0MmUwOGJjNTg3YWFjY2Q4IiwidHlwIjoiSldUIn0.eyJuYmYiOjE1MjM1NTQ3OTQsImV4cCI6MTUyMzU1ODM5NCwiaXNzIjoiaHR0cDovL2xvY2FsaG9zdDo1MzcyNyIsImF1ZCI6WyJodHRwOi8vbG9jYWxob3N0OjUzNzI3L3Jlc291cmNlcyIsImZoaXItYXBpIl0sImNsaWVudF9pZCI6Imtub3duLWNsaWVudC1pZCIsInNjb3BlIjpbImZoaXItYXBpIl19.pZWIWy3RdDHp5zgcYs8bb9VrxIHXbYu8LolC3YTy6xWsPxMoPUQwbAltYmC6WDXFiDygpsC5ofkGlR4BH0Bt1FMvFWqFYhPcOOKvBqLLc055EHZfTcNcmiUUf4y4KRuQFqWZsH_HrfWwykSGVio2OnYcQvytrbjAi_EzHf2vrHJUHX2JFY4A_F6WpJbQiI1hUVEOd7h1jfmAptWlNGwNRbCF2Wd1Hf_Hodym8mEOKQz21VHdvNJ_B-owPMvLjalV5Nrvpv0yC9Ly5YablrkzB583eHwQNSA7A4ZMm49O8MWv8kUwwF5TF0lJJDyyw3ruqmPWCM-058chenU0rtCsPQ
```

## Further Work Required
- Add different role requirements to different endpoints.
- Test this works with AAD
